File: app/src/ingestion/eventHandlers.js
const postgres = require('../databases/postgres');
const mongo = require('../databases/mongo');
const neo4j = require('../databases/neo4j');
const retryQueue = require('../utils/retryQueue');
const logger = require('../utils/logger');

// ─── DRIVER_LOCATION_UPDATE ───────────────────────────────────────────────────

async function handleDriverLocationUpdate(event) {
  const { driver_id, latitude, longitude, zone_id, timestamp, event_id } = event;

  if (!driver_id || !zone_id) {
    logger.error('DRIVER_LOCATION_UPDATE missing required fields', { event_id });
    return;
  }

  const cypher = `
    MERGE (d:Driver {driverId: $driverId})
    SET d.latitude = $latitude,
        d.longitude = $longitude,
        d.lastSeen = $timestamp,
        d.eventId = $eventId
    MERGE (z:Zone {zoneId: $zoneId})
    MERGE (d)-[r:LOCATED_IN]->(z)
    SET r.timestamp = $timestamp
    RETURN d, z
  `;

  try {
    await neo4j.runQuery(cypher, {
      driverId: driver_id,
      latitude,
      longitude,
      timestamp,
      eventId: event_id,
      zoneId: zone_id,
    });
    logger.info('Driver location updated in Neo4j', { driver_id, zone_id });
  } catch (err) {
    logger.error('Failed to update driver location in Neo4j', {
      driver_id,
      error: err.message,
    });
    throw err;
  }
}

// ─── PACKAGE_STATUS_CHANGE ────────────────────────────────────────────────────

async function handlePackageStatusChange(event) {
  const { package_id, status, timestamp, driver_id, location, event_id } = event;

  if (!package_id || !status) {
    logger.error('PACKAGE_STATUS_CHANGE missing required fields', { event_id });
    return;
  }

  const statusEntry = {
    status,
    timestamp,
    driver_id,
    location,
    event_id,
  };

  try {
    const collection = await mongo.getCollection('packages');

    await collection.updateOne(
      { package_id },
      {
        $set: { package_id, updated_at: timestamp },
        $push: {
          status_history: {
            $each: [statusEntry],
            $sort: { timestamp: 1 },
          },
        },
      },
      { upsert: true }
    );

    logger.info('Package status updated in MongoDB', { package_id, status });
  } catch (err) {
    logger.error('Failed to update package status in MongoDB', {
      package_id,
      error: err.message,
    });
    throw err;
  }
}

// ─── BILLING_EVENT ────────────────────────────────────────────────────────────

async function handleBillingEvent(event) {
  const { invoice_id, package_id, amount, customer_id, currency, timestamp, event_id } = event;

  if (!invoice_id || !package_id) {
    logger.error('BILLING_EVENT missing required fields', { event_id });
    return;
  }

  // Check if package is delivered in MongoDB
  const isDelivered = await checkPackageDelivered(package_id);

  if (!isDelivered) {
    logger.warn('Package not yet DELIVERED, deferring billing event to retry queue', {
      package_id,
      invoice_id,
    });
    retryQueue.enqueue(event);
    return;
  }

  await insertInvoice({ invoice_id, package_id, amount, customer_id, currency, timestamp, event_id });
}

async function checkPackageDelivered(packageId) {
  try {
    const collection = await mongo.getCollection('packages');
    const doc = await collection.findOne({ package_id: packageId });

    if (!doc || !doc.status_history || doc.status_history.length === 0) {
      return false;
    }

    // Check the latest status
    const sorted = [...doc.status_history].sort(
      (a, b) => new Date(b.timestamp) - new Date(a.timestamp)
    );
    return sorted[0].status === 'DELIVERED';
  } catch (err) {
    logger.error('Failed to check package delivery status', { packageId, error: err.message });
    return false;
  }
}

async function insertInvoice({ invoice_id, package_id, amount, customer_id, currency, timestamp, event_id }) {
  try {
    await postgres.query(
      `INSERT INTO invoices (invoice_id, package_id, amount, customer_id, currency, timestamp, event_id)
       VALUES ($1, $2, $3, $4, $5, $6, $7)`,
      [invoice_id, package_id, amount, customer_id, currency || 'USD', timestamp, event_id]
    );
    logger.info('Invoice inserted into PostgreSQL', { invoice_id, package_id });
  } catch (err) {
    if (err.code === '23505') {
      // Unique violation
      logger.error('Duplicate invoice_id detected, skipping insertion', {
        invoice_id,
        package_id,
        error: err.message,
      });
    } else {
      logger.error('Failed to insert invoice into PostgreSQL', {
        invoice_id,
        error: err.message,
      });
      throw err;
    }
  }
}

module.exports = {
  handleDriverLocationUpdate,
  handlePackageStatusChange,
  handleBillingEvent,
  checkPackageDelivered,
  insertInvoice,
};

File: app/src/ingestion/event_handlers.py
from src.databases import postgres, mongo, neo4j
from src.utils import retry_queue
from src.utils.logger import get_logger

logger = get_logger(__name__)


def handle_driver_location_update(event: dict):
    driver_id = event.get("driver_id")
    zone_id = event.get("zone_id")
    latitude = event.get("latitude")
    longitude = event.get("longitude")
    timestamp = event.get("timestamp")
    event_id = event.get("event_id")

    if not driver_id or not zone_id:
        logger.error(f"DRIVER_LOCATION_UPDATE missing required fields: event_id={event_id}")
        return

    cypher = """
        MERGE (d:Driver {driverId: $driverId})
        SET d.latitude = $latitude,
            d.longitude = $longitude,
            d.lastSeen = $timestamp,
            d.eventId = $eventId
        MERGE (z:Zone {zoneId: $zoneId})
        MERGE (d)-[r:LOCATED_IN]->(z)
        SET r.timestamp = $timestamp
        RETURN d, z
    """
    try:
        neo4j.run_query(cypher, {
            "driverId": driver_id,
            "latitude": latitude,
            "longitude": longitude,
            "timestamp": timestamp,
            "eventId": event_id,
            "zoneId": zone_id,
        })
        logger.info(f"Driver location updated in Neo4j: driver_id={driver_id} zone_id={zone_id}")
    except Exception as e:
        logger.error(f"Failed to update driver location in Neo4j: driver_id={driver_id} error={e}")
        raise


def handle_package_status_change(event: dict):
    package_id = event.get("package_id")
    status = event.get("status")
    timestamp = event.get("timestamp")
    driver_id = event.get("driver_id")
    location = event.get("location")
    event_id = event.get("event_id")

    if not package_id or not status:
        logger.error(f"PACKAGE_STATUS_CHANGE missing required fields: event_id={event_id}")
        return

    status_entry = {
        "status": status,
        "timestamp": timestamp,
        "driver_id": driver_id,
        "location": location,
        "event_id": event_id,
    }

    try:
        collection = mongo.get_collection("packages")
        collection.update_one(
            {"package_id": package_id},
            {
                "$set": {"package_id": package_id, "updated_at": timestamp},
                "$push": {"status_history": status_entry},
            },
            upsert=True,
        )
        logger.info(f"Package status updated in MongoDB: package_id={package_id} status={status}")
    except Exception as e:
        logger.error(f"Failed to update package status in MongoDB: package_id={package_id} error={e}")
        raise


def check_package_delivered(package_id: str) -> bool:
    try:
        collection = mongo.get_collection("packages")
        doc = collection.find_one({"package_id": package_id})
        if not doc or not doc.get("status_history"):
            return False
        sorted_history = sorted(doc["status_history"], key=lambda x: x.get("timestamp", ""), reverse=True)
        return sorted_history[0].get("status") == "DELIVERED"
    except Exception as e:
        logger.error(f"Failed to check package delivery status: package_id={package_id} error={e}")
        return False


def insert_invoice(event: dict):
    invoice_id = event.get("invoice_id")
    package_id = event.get("package_id")
    amount = event.get("amount")
    customer_id = event.get("customer_id")
    currency = event.get("currency", "USD")
    timestamp = event.get("timestamp")
    event_id = event.get("event_id")

    try:
        postgres.execute(
            """INSERT INTO invoices (invoice_id, package_id, amount, customer_id, currency, timestamp, event_id)
               VALUES (%s, %s, %s, %s, %s, %s, %s)""",
            (invoice_id, package_id, amount, customer_id, currency, timestamp, event_id),
        )
        logger.info(f"Invoice inserted into PostgreSQL: invoice_id={invoice_id} package_id={package_id}")
    except Exception as e:
        if "unique" in str(e).lower() or "duplicate" in str(e).lower() or "23505" in str(e):
            logger.error(f"Duplicate invoice_id detected, skipping insertion: invoice_id={invoice_id} error={e}")
        else:
            logger.error(f"Failed to insert invoice: invoice_id={invoice_id} error={e}")
            raise


def handle_billing_event(event: dict):
    invoice_id = event.get("invoice_id")
    package_id = event.get("package_id")

    if not invoice_id or not package_id:
        logger.error(f"BILLING_EVENT missing required fields: event_id={event.get('event_id')}")
        return

    is_delivered = check_package_delivered(package_id)

    if not is_delivered:
        logger.warning(f"Package not yet DELIVERED, deferring billing event: package_id={package_id} invoice_id={invoice_id}")
        retry_queue.enqueue(event)
        return

    insert_invoice(event)
